keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -280,13 +279,21 @@ _This section must be completed when targeting beta graduation to a release._
   an object specifying the range and validating if the traffic is allowed within 
   the specified range
 
+* **How can someone using this feature know that it is working for their instance?
+  - [x] Other
+   - Details:
+      The API Field must be present when a NetworkPolicy is created with that field.
+      The feature working correctly depends on the CNI implementation, so it can be assessed
+      by the operator verifying if the data path contains the correct network filtering rules 
+      and also testing if communications works as desired.
```

Comment:
Testing communications won't work for operators if they have a number of network policies per cluster and hundreds or more of clusters.
We should really try to make a metric-based.

> it can be assessed by the operator verifying if the data path contains the correct network filtering rules

Can we maybe ask CNI to expose some metric for it?

Reply:
well this is a hard subject that we usually discuss. CNIs usually can provide metrics, but they work on a conversion behavior other than only relying on the api object.

For example, Calico have a library that converts k8s API to internal structures (Calico API), so we would need to ask for them to provide metrics of that specific field, which can be pretty hard.

If this is a no-go for the feature, I can discuss with some of them, but IMO I guess that if we ask this to the CNIs we might be crossing Kubernetes boundaries. 

Anyway, we usually make the Network Policy e2e tests using a Calico installation, and there's already some work in progress to support the EndPort field in e2e tests, which might be a sign the feature working correctly or not.

We still need to update Calico version from the job to v3.19 (the version that supports EndPort) and enable the feature gate. I think we can do this as a pre-req before moving to stable.

@jayunit100 @knabben thoughts here?

Reply:
I don't actually see a status on `NetworkPolicy`.  How does a cluster-admin know that his other changes to NetworkPolicy have taken effect?

Reply:
> I don't actually see a status on NetworkPolicy. How does a cluster-admin know that his other changes to NetworkPolicy have taken effect?

That's a good point - but I don't think it's strictly related to this KEP...


> We still need to update Calico version from the job to v3.19 (the version that supports EndPort) and enable the feature gate. I think we can do this as a pre-req before moving to stable.

I will try to discuss a bit more within PRR team, but I think some mechanism (other than try yourself and see if it works) is really needed.

Reply:
We could propose adding a status. It would be up to CNI implementations to support it, of course. But with that, at least we would know that policies have been applied successfully (or not).

Reply:
Well, about Status in NetworkPolicy I think it can be something useful and necessary, although coordinating CNIs feeding events/status with rules status might be hard (as the rule can be created in one node, in multiple nodes depending on Pod spread, etc).

There's another [KEP](https://github.com/kubernetes/enhancements/pull/2137) with this sort of discussion, mainly about CNIs providing feedbacks on Netpol features, etc.

We're also discussing about a new Network Policy specification (v2, and ClusterNetworkPolicy) so maybe we can shift the Status discussion to the evolution of the NetPol API.

@astoycos I guess you might be interested in this discussion :)

Reply:
Yeah - we definitely don't want to block this KEP on adding status to NetworkPolicy.

That said - to unblock this KEP, I would like to at least see:
- more detailed description on how we can discover/debug potential problems
- at least a hand-wavy description on how we would improve it later (and probably block GA on that?)

Reply:
@wojtek-t so, we've been using a side project called Cyclonus (by @mattfenwick ) to check some conformance of CNIs. Basically it's a network policy fuzzer that creates a bunch of Network Policies and verifies if the result is as expected.

There's some more details here: https://kubernetes.io/blog/2021/04/20/defining-networkpolicy-conformance-cni-providers/

But we can, for instance create a periodic job that runs with Calico, for example to check potential problems. If this sounds good, I can add this to the KEP, WDYT?

Reply:
Happy to chat about ways we can apply Cyclonus to derisking this change!

Reply:
Adding a E2E job upstream and upgrade Calico CNI for a version to supports endport in parallel with cyclonus, sounds a good plan for covering the feature. +1

Reply:
This looks great, but let me reiterate one more time.
I'm not really asking for more tests (those are also desired, but it's not what we need here).
Some bugs will always happen.

So I'm asking here "how, as an operator of a cluster, I can say if my clusters work fine wrt this feature".
If I operate a hundred or thousand clusters, I obviously can't check the rules in all my clusters.
I'm also not interested in my own workload (which may work fine, but other workload may not).
I'm asking - how I can say that a given cluster works fine.

Reply:
@wojtek-t I can't see today a way of doing this other than with adding some Status.

CNIs have metrics (Calico, for instance have metrics of rules applied, errors when applying rules) so we can use those metrics. We can say, for example 'I know my cluster is working because I have rules applied, but my applying error rate has increased after I've enabled this feature'.

Is this something we want to add here? :)

Reply:
> @wojtek-t I can't see today a way of doing this other than with adding some Status.

I don't want to block Beta, but I would say that's something we should consider a blocker for GA.

> CNIs have metrics (Calico, for instance have metrics of rules applied, errors when applying rules) so we can use those metrics. We can say, for example 'I know my cluster is working because I have rules applied, but my applying error rate has increased after I've enabled this feature'.

Yes - this is exactly what I'm looking for.
If you would write that e.g. Calico has metrics foo and bar and some other CNI has metrics bar and baz - that would be perfect.